SOURCES/etc/xapi.d/plugins/smartctl.py

```diff
@@ -10,36 +10,39 @@
 from xcpngutils.operationlocker import OperationLocker
 
 @error_wrapped
-def _list_disks():
-    disks = []
+def _list_devices():
+    devices = []
     result = run_command(['smartctl', '--scan'])
     for line in result['stdout'].splitlines():
-        if line.startswith('/dev/') and not line.startswith('/dev/bus/'):
-            disks.append(line.split()[0])
-    return disks
+        devices.append({'name': line.split()[0], 'type':line.split()[2]})
+    return devices
 
 @error_wrapped
 def get_information(session, args):
     results = {}
+    i = 0
     with OperationLocker():
-        disks = _list_disks()
-        for disk in disks:
-            cmd = run_command(["smartctl", "-j", "-a", disk], check=False)
-            results[disk] = json.loads(cmd['stdout'])
+        devices = _list_devices()
+        for device in devices:
+            cmd = run_command(["smartctl", "-j", "-a", "-d", devices[i]['name'], devices[i]['type']], check=False)
+            results[device] = json.loads(cmd['stdout'])
+            i = i + 1
         return json.dumps(results)
 
 @error_wrapped
 def get_health(session, args):
     results = {}
+    i = 0
     with OperationLocker():
-        disks = _list_disks()
-        for disk in disks:
-            cmd = run_command(["smartctl", "-j", "-H", disk])
+        devices = _list_devices()
+        for device in devices:
+            cmd = run_command(["smartctl", "-j", "-H", "-d", devices[i]['name'], devices[i]['type']])
```

Comment:
Also I understand why it fails on my machine. It is because the exit status of the command is 4. In you case I think that it is 0 so it is ok. To fix my machine you just need to add the parameter `check=False) at the end of the call. Like this: `cmd = run_command(["smartctl", "-j", "-H", "-d", device['name'], device['type']], check=False)`

Reply:
Do you think that you can give a try using this one [smartctl.py](https://github.com/xcp-ng/xcp-ng-xapi-plugins/blob/gtn-pr-46/SOURCES/etc/xapi.d/plugins/smartctl.py) ?